### PR TITLE
Allow Non-Incremental `vllm.RegexLogitsProcessor` (enable fast-forward)

### LINKED ISSUE #855: Logits Processors `Guide` integration will be buggy when `len(tokens) > 1` in a `Write` instruction
### Describe the issue as clearly as possible:

See:
https://github.com/outlines-dev/outlines/blob/d6a2b7908065d420456118723f69908c4094c1f8/outlines/integrations/vllm.py#L110

Here the `tokens` field of the next instruction is treated equally regardless of whether it is of type `Generate` or `Write`. 

If a `Write` instruction has a `tokens` field with length > 1, this means we will accept any of the next ff-tokens as the token in the next step. This is incorrect.

### Steps/code to reproduce the bug:

```python
The bug will only appear once the ff-tokens have length > 1.
```

### Expected result:

```shell
N.a.
```

### Error message:

_No response_

### Outlines/Python version information:

Version information
<details>

```
python -c "import sys; print('Python', sys.version)"
pip freeze
0.0.39
Python 3.8.18 (default, Oct  2 2023, 15:02:11) 
[GCC 9.4.0]
accelerate==0.27.2
ai2-olmo==0.2.5
aiohttp==3.9.3
aioprometheus==23.12.0
aiosignal==1.3.1
annotated-types==0.6.0
antlr4-python3-runtime==4.9.3
anyio==4.2.0
async-timeout==4.0.3
attrs==23.2.0
awscli==1.32.83
boto3==1.34.83
botocore==1.34.83
cached_path==1.6.2
cachetools==5.3.3
certifi==2024.2.2
charset-normalizer==3.3.2
click==8.1.7
cloudpickle==3.0.0
cmake==3.29.2
codespell==2.2.6
colorama==0.4.4
cupy-cuda12x==12.1.0
diskcache==5.6.3
distro==1.9.0
docutils==0.16
einops==0.7.0
exceptiongroup==1.2.0
fastapi==0.109.2
fastrlock==0.8.2
filelock==3.13.1
frozenlist==1.4.1
fsspec==2024.2.0
google-api-core==2.18.0
google-auth==2.29.0
google-cloud-core==2.4.1
google-cloud-storage==2.16.0
google-crc32c==1.5.0
google-resumable-media==2.7.0
googleapis-common-protos==1.63.0
h11==0.14.0
hiredis==2.3.2
httpcore==1.0.4
httptools==0.6.1
httpx==0.27.0
huggingface-hub==0.20.3
idna==3.6
importlib-resources==6.1.1
importlib_metadata==7.0.2
iniconfig==2.0.0
interegular==0.3.3
isort==5.13.2
Jinja2==3.1.3
jmespath==1.0.1
joblib==1.3.2
jsonschema==4.21.1
jsonschema-specifications==2023.12.1
lark==1.1.9
libnacl==2.1.0
llvmlite==0.41.1
lm-format-enforcer==0.9.3
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
mpmath==1.3.0
msgpack==1.0.7
multidict==6.0.5
mypy==1.9.0
mypy-extensions==1.0.0
nest-asyncio==1.6.0
networkx==3.1
ninja==1.11.1.1
numba==0.58.1
numpy==1.24.4
nvidia-cublas-cu12==12.1.3.1
nvidia-cuda-cupti-cu12==12.1.105
nvidia-cuda-nvrtc-cu12==12.1.105
nvidia-cuda-runtime-cu12==12.1.105
nvidia-cudnn-cu12==8.9.2.26
nvidia-cufft-cu12==11.0.2.54
nvidia-curand-cu12==10.3.2.106
nvidia-cusolver-cu12==11.4.5.107
nvidia-cusparse-cu12==12.1.0.106
nvidia-nccl-cu12==2.18.1
nvidia-nvjitlink-cu12==12.3.101
nvidia-nvtx-cu12==12.1.105
omegaconf==2.3.0
openai==1.13.3
orjson==3.9.13
outlines==0.0.39
packaging==23.2
peft==0.9.0
pillow==10.3.0
pkgutil_resolve_name==1.3.10
pluggy==1.4.0
prometheus_client==0.20.0
proto-plus==1.23.0
protobuf==4.25.2
psutil==5.9.8
py==1.11.0
py-cpuinfo==9.0.0
pyasn1==0.6.0
pyasn1_modules==0.4.0
pydantic==2.6.1
pydantic_core==2.16.2
Pygments==2.17.2
pynvml==11.5.0
pytest==8.0.2
pytest-asyncio==0.23.5
pytest-forked==1.6.0
pytest-rerunfailures==13.0
pytest-shard==0.1.2
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
PyYAML==6.0.1
quantile-python==1.1
ray==2.9.2
redis==5.0.3
referencing==0.33.0
regex==2023.12.25
requests==2.31.0
rich==13.7.1
rpds-py==0.17.1
rsa==4.7.2
ruff==0.1.5
s3transfer==0.10.1
safetensors==0.4.2
scipy==1.10.1
sentencepiece==0.1.99
six==1.16.0
sniffio==1.3.0
starlette==0.36.3
sympy==1.12
tensorizer==2.9.0a0
tiktoken==0.6.0
tokenizers==0.19.1
toml==0.10.2
tomli==2.0.1
torch==2.1.2
tqdm==4.66.1
transformers==4.40.0
triton==2.1.0
types-PyYAML==6.0.12.12
types-requests==2.31.0.6
types-setuptools==69.1.0.20240308
types-urllib3==1.26.25.14
typing_extensions==4.9.0
urllib3==1.26.18
uvicorn==0.27.0.post1
uvloop==0.19.0
-e git+ssh://git@github.com/br3no/vllm.git@e6ffb1af2e904436473568f9d43131c998e55063#egg=vllm
watchfiles==0.21.0
websockets==12.0
xformers==0.0.23.post1
yapf==0.32.0
yarl==1.9.4
zipp==3.17.0
```

</details>

### Context for the issue:

Bug was discussed in a call with @rlouf.

### PR DESCRIPTION
**Pull request by lapp0**
Fixes #855 

## Problem:

If multiple new tokens are introduced to `vllm.RegexLogitsProcessor` without the previous sequence (`input_ids[:-1]`) being processed, `state` defaults to `0` in

```
self._fsm_state[seq_id] = self.fsm.get_next_state(
    state=self._fsm_state[last_seq_id],   # DefaultDict, if last_seq_id doesn't exist it assumes initial state 0!
    token_id=last_token
)
```

Example of problematic calls:
```
lp = RegexLogitsProcessor(...)

biased_scores = lp([], scores)
biased_scores = lp([1], scores)
biased_scores = lp([1, 2, 3, 4, 5], scores) 
```

## Solution

Calculate all missing FSM states sequentially, enabling fast-forward.

Would appreciate a review @br3no, I'd like ensure this is compliant with expectations within vLLM :)

### CODE REVIEW

**Comment by br3no** [comment]
@lapp0 I think there is a misunderstanding here. 

In #855 the issue has to do with new type `Instruction = Union[Write, Generate]` in https://github.com/outlines-dev/outlines/blob/main/outlines/fsm/guide.py and the usage of the `get_next_instruction` method of `Guide` in the `LogitsProcessor`. The thing is that both instructions `Write` and `Generate` declare a member variable called `tokens` but the semantics of this member variable is not the same. 

In `Write`, `tokens` means the sequence of next tokens that Outlines wants to "write" to the sequence. These are the ff-tokens I was talking about. `tokens` is to be interpreted as the sorted list of next tokens for the sequence being generated.

In `Generate`, `tokens` means the set of tokens that Outlines says are valid (i.e. transitions in the state machine) for the next generation step. Only one of these tokens will eventually be generated by the LLM.

You can see how I will make sure the bug doesn't happen in vLLM, once the update of Outlines is merged here: https://github.com/br3no/vllm/blob/e6ffb1af2e904436473568f9d43131c998e55063/vllm/model_executor/guided_decoding/outlines_logits_processors.py#L59

Note that this way of solving the issue makes using ff-tokens from outlines impossible... So it can only be considered an interim solution.

---

Regarding the integration with vLLM: there is ongoing work in vLLM to allow for ff-tokens and the design of how they will work is still unclear (at least to me). So I'm not sure the `LogitsProcessor` interface will change in the way you assume here, i.e. that the `input_ids` parameter to `__call__` could contain `n` (n > 1) newly generated tokens.

**Comment by lapp0** [comment]
Thanks so much for your detailed response @br3no 

Here are my thoughts:

0) Current PR changes should remain to ensure if multiple tokens are added to `input_ids` the logits processor can traverse the automata correctly.

1) The fact that `Write.tokens` and `Generate.tokens` means separate things is confusing. IMO we should change it to `Write.tokens` and `Generate.legal_next_tokens`.

2) Application of `Write` instructions is out of scope for a logits processor. We probably need two separate components in vLLM: a guide logits processor, and a guide fast forward handler. 

You have more context so I'd love to hear any corrections / guidance / thoughts. But here's what I'm thinking: The logits processor should behave as it does now: it creates a mask with a single token provided a `Write` instruction.  The fast forward handler would share the `Guide` object with the logits processor. The fast forward handler would return an empty sequence upon Generate instructions, and `Write.tokens` upon a write instruction. Infrastructure for `Speculative` might be applicable for the fast forward handler.

What are your thoughts on the direction we should take and how can I help?

**Comment by br3no** [comment]
I'm not sure there ever will be multiple tokens added to `input_ids`, so I'd rather wait for vLLM to define the way things are supposed to work before implementing this.

> 1. ...
Yes, I agree there is room for reducing confusion. Maybe you want to propose the change in a separate PR and discuss this with the Outlines maintainers?

Regarding the `LogitsProcessor`'s API, this is something that is being worked on in vLLM. I think there is not much one can do right now but to wait what will come out of this work.

**Comment by lapp0** [comment]
Thanks for the clarification @br3no!
